torchao/quantization/qat/api.py

```diff
@@ -5,25 +5,230 @@
+    not exist yet.
+
+    Example usage::
+
+        from torchao.quantization import quantize_
+        from torchao.quantization.qat import IntxFakeQuantizeConfig
+
+        activation_config = IntxFakeQuantizeConfig(
+            torch.int8, "per_token", is_symmetric=False,
+        )
+        weight_config = IntxFakeQuantizeConfig(
+            torch.int4, group_size=32, is_symmetric=True,
+        )
+        qat_config = QATConfig(
+            # must specify one of `base_config` or `weight_config`
+            activation_config=act_config,
+            weight_config=weight_config,
+            step="prepare",
+        )
+        quantize_(model, qat_config)
+
+    Args:
+        base_config (Optional[AOBaseConfig]): Base PTQ config to infer the fake
+            quantization configs during the prepare phase, and to apply directly
+            during the convert phase.
+        activation_config (Optional[FakeQuantizeConfigBase]): Custom fake
+            quantization config for input activations, always optional.
+            Must be None if `base_config` is used.
+        weight_config (Optional[FakeQuantizeConfigBase]): Custom fake quantization
+            config for weights. Must be None if `base_config` is used.
+
+    Keyword args:
+        step (str): One of "prepare" or "convert", determines the QAT phase
+
+    Raises:
+        ValueError: If `base_config` and `activation_config` are both specified
+        ValueError: If `base_config` and `weight_config` are both specified
+        ValueError: If neither `base_config` nor `weight_config` is specified
+        ValueError: If `step` is not one of "prepare" or "convert"
+        ValueError: If `base_config` is None but `step` is "convert"
+        ValueError: If the config is applied on a module that is not a
+            `torch.nn.Linear` or `torch.nn.Embedding`, or it is applied on
+            `torch.nn.Embedding` with an activation config
+    """
+
+    base_config: Optional[AOBaseConfig]
+    activation_config: Optional[FakeQuantizeConfigBase]
+    weight_config: Optional[FakeQuantizeConfigBase]
+    step: QATConfigStep
+
+    # Express `step` as a keyword argument
+    # TODO: Use `kw_only=True` instead, added in python 3.10
+    def __init__(
+        self,
+        base_config: Optional[AOBaseConfig] = None,
+        activation_config: Optional[FakeQuantizeConfigBase] = None,
+        weight_config: Optional[FakeQuantizeConfigBase] = None,
+        *,
+        step: QATConfigStep = "prepare",
+    ):
+        self.base_config = base_config
+        self.activation_config = activation_config
+        self.weight_config = weight_config
+        self.step = step
+        self.__post_init__()
+
+    def __post_init__(self):
+        self.step = self.step.lower()
+        all_step_values = [s.value for s in QATConfigStep]
+        if self.step not in all_step_values:
+            raise ValueError("`step` must be one of %s" % all_step_values)
```

Comment:
nit: this can be removed now? will it be cause by the enum itself?

Reply:
seems like it's still needed: https://github.com/pytorch/ao/blob/551703147e050e8770c411019a7230de87c20e72/test/quantization/test_qat.py#L1288